### PR TITLE
add the ability to compile models inside dbt packages

### PR DESCRIPTION
**Pull request by pgrivachev**
Now it is possible to open model file from the dbt_packages folder and to see the compilation results.

### CODE REVIEW

File: server/src/DbtTextDocument.ts [hunk 1]
```diff
@@ -203,6 +203,14 @@ export class DbtTextDocument {
 
   getModelPath(): string {
     const index = this.rawDocument.uri.indexOf(this.workspaceFolder);
+    const filePath = this.rawDocument.uri.slice(index + this.workspaceFolder.length + 1);
+    if (filePath.startsWith('dbt_packages') || filePath.startsWith('dbt_modules')) {
+      return filePath
+        .replaceAll('/', '.')
+        .replace(/^(dbt_packages|dbt_modules)./, '')
+        .replace('models.', '')
```
Comment (fivetran-dmitryboykov) [inline server/src/DbtTextDocument.ts:211]: Models location could be overwritten in `dbt_project.yml`. Moreover models could be located in multiple directories. E.g. `source-paths: ["models", "other_models"]`.

File: e2e/projects/project-with-packages/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+dbt_packages
```
Comment (fivetran-dmitryboykov) [inline e2e/projects/project-with-packages/.gitignore:1]: Please add `dbt_modules`.

File: server/src/DbtTextDocument.ts (hunk 1 above)
Comment (fivetran-dmitryboykov) [reply server/src/DbtTextDocument.ts:211]: Let's think if it would be better to create separate utils function, e.g. `extractModelName`.

Comment (fivetran-dmitryboykov) [reply server/src/DbtTextDocument.ts:211]: `dbt_packages` and `dbt_modules` to consts?

File: client/src/ExtensionClient.ts [hunk 3]
```diff
@@ -187,24 +187,28 @@ export class ExtensionClient {
...
     const outerWorkspace = this.workspaceHelper.getOuterMostWorkspaceFolder(folder);
 
-    let currentUri = uri;
+    let currentUri = fileUri;
     do {
       currentUri = Uri.joinPath(currentUri, '..');
       try {
         await workspace.fs.stat(currentUri.with({ path: `${currentUri.path}/dbt_project.yml` }));
+        const oneLevelUpPath = Uri.joinPath(currentUri, '..').path;
+        if (oneLevelUpPath.endsWith('dbt_packages') || oneLevelUpPath.endsWith('dbt_modules')) {
```
Comment (fivetran-dmitryboykov) [inline client/src/ExtensionClient.ts:209]: `dbt_packages` and `dbt_modules` to consts?

**Review by fivetran-dmitryboykov — CHANGES_REQUESTED** [review]
Generally looks good. But which features will be available for dbt package models ?
Go to definition isn't working - we need to use proper package name for such models.

**Comment by pgrivachev** [comment]
> Generally looks good. But which features will be available for dbt package models ? Go to definition isn't working - we need to use proper package name for such models.

Good point! Looks like I can fix it.

**Review by fivetran-dmitryboykov — CHANGES_REQUESTED** [review]
Looks good. Just some questions.

File: server/src/DbtTextDocument.ts [hunk 4]
```diff
@@ -200,17 +200,44 @@ export class DbtTextDocument {
...
+  static getFilePathRelatedToWorkspace(docUri: string, workspaceFolder: string): string {
+    const index = docUri.indexOf(workspaceFolder);
+    return docUri.slice(index + workspaceFolder.length + 1);
+  }
+
+  static getModelPathOrFullyQualifiedName(docUri: string, workspaceFolder: string, dbtRepository: DbtRepository): string {
+    const filePath = this.getFilePathRelatedToWorkspace(docUri, workspaceFolder);
+    if (dbtRepository.packagesInstallPaths.some(p => filePath.startsWith(p))) {
+      const startWithPackagesFolder = new RegExp(`^(${dbtRepository.packagesInstallPaths.join('|')}).`);
+      const modelsFolder = new RegExp(`(${dbtRepository.modelPaths.join('|')}).`);
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/DbtTextDocument.ts:219]: Oh yes, you're right, I think for now we can just assume that the models in the packages are located in the models folder.

File: server/src/DbtRepository.ts [hunk 5]
```diff
@@ -1,9 +1,25 @@
 import { ManifestMacro, ManifestModel, ManifestSource } from './manifest/ManifestJson';
 
 export class DbtRepository {
+  static readonly DBT_PROJECT_FILE_NAME = 'dbt_project.yml';
+  static readonly DBT_MANIFEST_FILE_NAME = 'manifest.json';
+  static readonly DBT_PROJECT_NAME_FIELD = 'name';
+
+  static readonly SOURCE_PATHS_FIELD = 'source-paths'; // v1.0.0: The config source-paths has been deprecated in favor of model-paths
+  static readonly MODEL_PATHS_FIELD = 'model-paths';
+  static readonly PACKAGES_INSTALL_PATH_FIELD = 'packages-install-path';
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/DbtRepository.ts]: Thank you, I missed this.

File: server/src/DbtTextDocument.ts [hunk 6]
```diff
@@ -200,17 +200,44 @@ export class DbtTextDocument {
...
+      return filePath.replaceAll('/', '.').replace(startWithPackagesFolder, '').replace(modelsFolder, '').replace(/.sql$/, '');
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/DbtTextDocument.ts:220]: Should be `package_name.examples.users`. At least it works for salesforce_source.tmp.stg_salesforce__user_tmp model in our test project.

**Pull request merged**